Refactor metaToJsonProperty to accept AJV keywords

iqbaldjulfri opened the pull request:
We need to add additional attributes (AJV keywords) from PropertyDefinition so they will be written to openapi.json and we can use it as validation. Currently, only type, description, items (for arrays), and $ref are written (and a hard-coded format: 'date-time' for Date instances).

issue #2684

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

jannyHou commented on packages/repository-json-schema/src/build-schema.ts:
```diff
@@ -108,15 +108,38 @@ export function metaToJsonProperty(meta: PropertyDefinition): JSONSchema {
+  // create a shallow copy of meta, so we can delete what we don't need
+  const metaCopy = {...meta};
+  // delete everything we don't need
+  [
```
How do you decide which properties are not needed?

iqbaldjulfri replied:
1. It is defined in the `PropertyDefinition`, but not in the AJV's keywords specs.
2. `type` & `description`, since they will be written by the original code.

raymondfeng commented:
@iqbaldjulfri Please fix the CI failures per instructions at https://loopback.io/doc/en/lb4/submitting_a_pr.html.

iqbaldjulfri commented:
@raymondfeng All CI errors fixed, I need to force push to change the earlier commit message.

iqbaldjulfri commented:
@bajtos thank you for the insight on the bigger picture of this problem. I've refactor the code to use `jsonSchema` as you suggested. With this we can leverage `ajv` validations in the OpenAPI layer.

bajtos commented on packages/repository-json-schema/src/__tests__/unit/build-schema.unit.ts:
```diff
@@ -169,5 +169,25 @@ describe('build-schema', () => {
+
+    it('keeps AJV keywords', () => {
+      expect(
+        metaToJsonProperty({
```
This statement is way too long to my taste, it also combines to parts of a test (act + assert) into a single line, which can make the test more difficult to understand. Let's split this into two statements please.

```ts
const schema = metaToJsonProperty({
  // ...
});

expect(schema).to.eql({
  // ...
});
```

iqbaldjulfri commented:
@bajtos, all done. I've implemented your suggestion for the unit test and squashed them into a single commit.

PS: Thank you guys for everything. This is my very first contribution to open source project and I'm very grateful to have you guiding a newcomer like me. Cheers!

bajtos commented:
Landed. 🎉 Thank you @iqbaldjulfri for the contribution! ❤️ 

> Thank you guys for everything. This is my very first contribution to open source project and I'm very grateful to have you guiding a newcomer like me.

You are very welcome 😄
